### PR TITLE
Update @automattic/calypso-build to v 5.0

### PR DESCRIPTION
**Pull request by ockham**
#### Changes proposed in this Pull Request:

Manual attempt at upgrading `@automattic/calypso-build` to v5, to surface potential bugs.
#### Is this a new feature or does it add/remove features to an existing part of Jetpack?

Upgrading a build tool.

#### Testing instructions:

```
yarn distclean && yarn
yarn build
yarn test-extensions
```

#### Proposed changelog entry for your changes:

Update @automattic/calypso-build to v5.

#### TODO

- [x] ~Currently fails to find PostCSS config.~ Fixed by https://github.com/Automattic/wp-calypso/pull/37110.
- [x] ~Currently doesn't transpile `import`s in config files.~ See comments for discussion.

### CODE REVIEW

**Comment by ockham** [comment]
With https://github.com/Automattic/wp-calypso/pull/37110 applied, this is still failing in `yarn build` stage. I've tracked down the issue and am working on a fix.

**Comment by ockham** [comment]
One possible fix is

```diff
diff --git a/babel.config.js b/babel.config.js
index ef8132aab..b3bcbb62e 100644
--- a/babel.config.js
+++ b/babel.config.js
@@ -5,6 +5,12 @@ const config = {
                        test: './extensions/',
                        presets: [ require.resolve( '@automattic/calypso-build/babel/wordpress-element' ) ],
                },
+               {
+                       test: [ './gulpfile.babel.js', './webpack.config.js', './tools/builder/' ],
+                       presets: [
+                               [ require.resolve( '@automattic/calypso-build/babel/default' ), { modules: 'commonjs' } ],
+                       ],
+               },
        ],
        env: {
                test: {

```

**Comment by ockham** [comment]
The simpler alternative is


```diff
diff --git a/package.json b/package.json
index d420c8c57..1287601a8 100644
--- a/package.json
+++ b/package.json
@@ -25,7 +25,7 @@
                "install-if-deps-outdated": "yarn check 2> /dev/null || yarn install --check-files --production=false",
                "distclean": "rm -rf node_modules && yarn clean",
                "build": "yarn install-if-deps-outdated && yarn clean && yarn build-client && yarn build-php && yarn build-extensions",
-               "build-client": "gulp",
+               "build-client": "MODULES=commonjs gulp",
                "build-extensions": "webpack --config ./webpack.config.extensions.js",
                "build-php": "composer install --ignore-platform-reqs",
                "build-production-php": "COMPOSER_MIRROR_PATH_REPOS=1 composer install -o --no-dev --classmap-authoritative --prefer-dist",
```

But I'm wondering if setting the env variable will cause all the code to be transpiled to CJS? (Though I'm not completely sure that that wouldn't be the case with https://github.com/Automattic/jetpack/pull/13854#issuecomment-546991707).

/cc @sirreal

**Comment by ockham** [comment]
I pushed 8cf9d13, which implements the more granular (per-file) option, which I'm pretty sure should only transpile `import` statements in the files listed.

The downside is that it needs to be manually maintained, if a new config file is added; I think it's reasonable to assume that that won't happen _that_ often.

I'm pretty sure the other option (setting the `MODULES` env var, i.e. `MODULES=commonjs gulp`) would cause ES Modules (`import` statements) in _all_ files to be transpiled to CommonJS (`require`).

**Comment by ockham** [comment]
`jest` is still failing to transpile `import`s in test files, which is weird, since we're [using](https://github.com/Automattic/jetpack/blob/975677f6c177859d672cc34851215ed35dcbda03/jest.config.js#L7) `@automattic/calypso-build`'s [Jest preset](https://github.com/Automattic/wp-calypso/blob/fba2d80dc733828ff3fff47af86f0944cca83bff/packages/calypso-build/jest-preset.js), which is [supposed to do just that](https://github.com/Automattic/wp-calypso/blob/fba2d80dc733828ff3fff47af86f0944cca83bff/packages/calypso-build/jest/transform/babel.js#L9) :confused:

File: tests/jest-extensions-babel-transform.js
Comment (sirreal) [inline tests/jest-extensions-babel-transform.js]: I believe the calypso-build jest preset is being overwritten. Rather than making any changes here (maybe this entire file can be removed?), let's try removing the transforms here:

https://github.com/Automattic/jetpack/blob/555b14071a8edf95a506ff51bb855f1a03a1232a/jest.config.js#L9-L14

Comment (ockham) [reply tests/jest-extensions-babel-transform.js]: Like so?

```diff
diff --git a/jest.config.js b/jest.config.js
index 10058a9d6..b17ab648b 100644
--- a/jest.config.js
+++ b/jest.config.js
@@ -6,10 +6,4 @@ const path = require( 'path' );
 module.exports = {
        preset: '@automattic/calypso-build',
        roots: [ '<rootDir>/extensions/' ],
-       transform: {
-               '\\.[jt]sx?$': path.join( __dirname, 'tests', 'jest-extensions-babel-transform' ),
-               '\\.(gif|jpg|jpeg|png|svg|scss|sass|css)$': require.resolve(
-                       '@automattic/calypso-build/jest/transform/asset'
-               ),
-       },
 };
```

That gives me two `ReferenceError: React is not defined` errors when running `npm run test-extensions` (i.e. `jest`), in `extensions/blocks/markdown/test/markdown-renderer.js` and `extensions/blocks/tiled-gallery/layout/mosaic/test/index.js` :confused:

Comment (ockham) [reply tests/jest-extensions-babel-transform.js]: It's because we're missing the [`wordpress-element` transform from `jest-extensions-babel-transform`](https://github.com/Automattic/jetpack/blob/555b14071a8edf95a506ff51bb855f1a03a1232a/tests/jest-extensions-babel-transform.js#L9).

Did I say that I hate `@wordpress/element`? :angry:

Comment (ockham) [reply tests/jest-extensions-babel-transform.js]: After a call with @sirreal, we decided to leave things as they are, since there's no straightforward way to simplify this.

**Comment by zinigor** [comment]
Tested the build, it seems to work as expected, plus it fixes the node-sass build error that breaks the build flow on Node 12, for example, in the GitHub action that we currently have for Jetpack.

**Comment by kraftbj** [comment]
JP Blocks are failing to load with a console error:

`TypeError: _wordpress_blocks__WEBPACK_IMPORTED_MODULE_2__ is undefined editor.js:21:1`

```
TypeError: _wordpress_blocks__WEBPACK_IMPORTED_MODULE_2__ is undefined editor.js:21:1
    <anonymous> webpack:///./extensions/shared/block-category.js?:21
    js https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:2547
    __webpack_require__ https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:72
    <anonymous> webpack:///./extensions/editor.js?:4
    js https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:2524
    __webpack_require__ https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:72
    <anonymous> webpack:///multi_./extensions/editor_./extensions/blocks/business-hours/editor.js_./extensions/blocks/contact-form/editor.js_./extensions/blocks/contact-info/editor.js_./extensions/blocks/gif/editor.js_./extensions/blocks/likes/editor.js_./extensions/blocks/mailchimp/editor.js_./extensions/blocks/map/editor.js_./extensions/blocks/markdown/editor.js_./extensions/blocks/publicize/editor.js_./extensions/blocks/recurring-payments/editor.js_./extensions/blocks/related-posts/editor.js_./extensions/blocks/repeat-visitor/editor.js_./extensions/blocks/sharing/editor.js_./extensions/blocks/shortlinks/editor.js_./extensions/blocks/simple-payments/editor.js_./extensions/blocks/slideshow/editor.js_./extensions/blocks/subscriptions/editor.js_./extensions/blocks/tiled-gallery/editor.js_./extensions/blocks/videopress/editor.js_./extensions/blocks/wordads/editor.js?:1
    0 https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:4525
    __webpack_require__ https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:72
    <anonymous> https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:249
    <anonymous> https://jn/wp-content/plugins/jetpack-dev/_inc/blocks/editor.js?ver=1572965765:252
```

**Comment by ockham** [comment]
The problem here is specifically that our blocks code cannot `import from @wordpress/blocks`. This is probably related to a change in `calypso-build` v4.2 that changes the default behavior to _not_ transpile ESM syntax (`import`) to CommonJS (`require`).

This transpilation isn't normally required for code that's meant to be run on the client side, since the bundler (webpack) will understand ESM syntax directly, and support features (such as tree shaking) that aren't available if ESM is transpiled to CommonJS.

However, there seems to be a problem with how such code interfaces with `external`ised code, i.e. code that's declared as being available/loaded outside of webpack's realm -- typically as a global variable (via a `<script />` tag). Webpack's `externals` setting supports various different options, [see](https://webpack.js.org/configuration/externals/).

This applies to our `@wordpress/` dependencies: They are available as `wp.` variables, and aliased to their `@wordpress/` counterpart syntax (e.g. `@wordpress/blocks` is resolved to `wp.blocks`, which is loaded by `blocks.min.js` via a `<script />` tag).

All of this is handled by `@wordpress/dependency-extraction-webpack-plugin`, which is used and configured by `calypso-build`. It appears that we need to update something (possibly one of ) there to make sure our untranspiled `import` statements continue to find code in our `external`ised modules.

Pinging `@wordpress/dependency-extraction-webpack-plugin` guru @sirreal :slightly_smiling_face:

**Comment by sirreal** [comment]
The `calypso-build` upgrade included an update to @wordpress/dependency-extraction-webpack-plugin@2.0.0, which has a breaking change where the dependency output format, name, and contents change. I added changes to this PR to bring these up to date.

Additionally, I used the updated plugin in the dashboard build and updated it accordingly, so they're consistent. That should fix the issues.

Test blocks, view and admin side, as well as the Jetpack admin dashboard. All now seem to be working as expected in my testing.

**Review by jeherve — DISMISSED** [review]
It seems we've lost the green color of the Jetpack icon for the block sidebar:

image

**Comment by ockham** [comment]
> It seems we've lost the green color of the Jetpack icon for the block sidebar: [...]

Thanks Jeremy, good spot. It seems like a few other colors are also affected; e.g. here's the Upgrade Nudge for the Recurring Payments block:

image

The reason is quite probably that `calypso-build` v5 no longer includes `calypso-color-schemes` but relies on users to include it with their projects. I'm gonna try to include it with Jetpack.

**Comment by ockham** [comment]
Specifically, we're using `--color-jetpack` here: 

https://github.com/Automattic/jetpack/blob/c9d1b64edc5022952cba22e217413176010e3c2b/extensions/shared/jetpack-plugin-sidebar.scss#L34

And it's coming from https://github.com/Automattic/wp-calypso/blob/ee8a00a728dfd620e406a7626b6264a1a4a518f6/packages/calypso-color-schemes/src/shared/color-schemes/_default.scss#L286

**Comment by ockham** [comment]
With 0faefe078, the styling for the Jetpack logo is as follows:

image

This is the PostCSS plugin at work, and it's doing its job correctly: It uses both the CSS var, and (as a fallback for legacy browsers), the 'computed' property. Browser standards mandate that the var take precedence; however, the var here (`--color-jetpack`) [depends on another var, `--studio-jetpack-green`](https://github.com/Automattic/wp-calypso/blob/ee8a00a728dfd620e406a7626b6264a1a4a518f6/packages/calypso-color-schemes/src/shared/color-schemes/_default.scss#L286) which isn't in `@automattic/calypso-color-schemes`, but in a separate npm, `@automattic/color-studio`. As a result, the Jetpack color still doesn't show as green :(

One obvious question that comes to mind is why doesn't `@automattic/calypso-color-schemes` work out-of-the-box, i.e. import/depend on definitions from `@automattic/color-studio`, since arguably a user of `@automattic/calypso-color-schemes` shouldn't have to worry about that. I don't have a very clear answer to this since I'm just not familiar enough with (modern) CSS techniques (aside, most of this _is_ actually CSS, not SCSS!); maybe it's really just not possible. /cc @porada for advice :)

Previous versions of `calypso-build` worked, since they included a very similar `postcss.config.js` as the one I just added to this PR, with one difference:

```diff
diff --git a/postcss.config.js b/postcss.config.js
index 731973e72..35c9aeb57 100644
--- a/postcss.config.js
+++ b/postcss.config.js
@@ -2,6 +2,7 @@ module.exports = () => ( {
        plugins: {
                'postcss-custom-properties': {
                        importFrom: [ require.resolve( '@automattic/calypso-color-schemes' ) ],
+                       preserve: false,
                },
                autoprefixer: {},
        },
```

This drops the `--var` based style definitions, and _only_ ships the fallback (which is obviously guaranteed to always work):

image

 This is obviously one solution that would get colors back for this PR; however, it's arguable that it's just a workaround to the underlying problem described above (and that we should ideally solve at `@automattic/calypso-color-schemes` or `@automattic/color-studio` level, since other projects will otherwise be bitten by the same issues).

/cc @jsnajdr and @Automattic/team-calypso for teh lulz

**Comment by ockham** [comment]
I also tried 

```diff
diff --git a/postcss.config.js b/postcss.config.js
index 731973e72..1de3cf006 100644
--- a/postcss.config.js
+++ b/postcss.config.js
@@ -1,7 +1,10 @@
 module.exports = () => ( {
        plugins: {
                'postcss-custom-properties': {
-                       importFrom: [ require.resolve( '@automattic/calypso-color-schemes' ) ],
+                       importFrom: [
+                               require.resolve( '@automattic/color-studio' ),
+                               require.resolve( '@automattic/calypso-color-schemes' ),
+                       ],
                },
```

to explicitly include definitions from `@automattic/color-studio` before `@automattic/calypso-color-schemes`, but that also didn't work. I'm going to push a commit using the `preserve: false` workaround to unblock this PR; let's continue discussion on how to properly solve the issue.

**Review by jeherve — DISMISSED** [review]
This seems to work in my tests, but I'd appreciate another pair of eyes on this before we hit merge :)

**Comment by ockham** [comment]
Rebased (since there was a conflict) :grimacing:

**Review by jeherve — APPROVED** [review]
Still looks good to me. 🚢

**Comment by kraftbj** [comment]
r199225-wpcom

**Comment by simison** [comment]
Looks like simple sites are missing the color still — if I understood correct, colors work in Jetpack dashboard just fine?


Screenshot 2019-11-22 at 13 07 12
Screenshot 2019-11-22 at 13 07 07

**Comment by simison** [comment]
Looks like `postcss.config.js` wasn't synced in D34594-code. Adding it to wpcom in D35821-code; I think that should fix all this for simple sites.

File: postcss.config.js
Comment (simison) [inline postcss.config.js]: I think it would make sense to use [`@automattic/color-studio`](https://github.com/Automattic/color-studio/) directly here.

I don't really see Calypso colors bringing any additional value; it just has some extra Calypso specific things such as `--sidebar-background`, `--masterbar-toggle-drafts-editor-background`, etc...

Unless I'm missing something?

Comment (ockham) [reply postcss.config.js]: Was considering that, yeah. I think we'd have to rename a few colors though, since they're aliased in `@automattic/calypso-color-schemes`.

Furthermore, I think there's still some issue that prevents things from working entirely as they should, see https://github.com/Automattic/jetpack/pull/13854#issuecomment-550898168 :thinking:

Comment (ockham) [reply postcss.config.js]: >  I think we'd have to rename a few colors though, since they're aliased in `@automattic/calypso-color-schemes`.

That might not have been very clear; what I meant is that we're right now using the names as found in `@automattic/calypso-color-schemes` in Jetpack (not names from `@automattic/color-studio`).

Comment (jsnajdr) [reply postcss.config.js]: > I don't really see Calypso colors bringing any additional value

`calypso-color-schemes` brings exactly the value that the package name promises: it defines a family of color schemes like Classic Blue, Midnight or Powder Snow that can be applied to UI that uses "semantic" names like `color-primary`, `color-accent` or `masterbar-item`.

`color-studio` is a generic set of named colors (somewhat similar to RAL or Pantone).

If you want to style your UI with a color scheme that was designed to work together, then you need one of the Calypso color schemes.

Comment (simison) [reply postcss.config.js]: Thanks for clarification!

Sounds like @Automattic/jetpack-design could have final say which one they prefer to work with?

Comment (folletto) [reply postcss.config.js]: I'm not entirely sure of the question. It seems a library choice question... what's the difference from a design perspective?

If it answer the question: going forward we'll use the latest `color-studio` colors, which includes the new Jetpack green and a gamut of colors agreed across all A8c.

Comment (jsnajdr) [reply postcss.config.js]: @folletto I think the question is more like: does Jetpack have something like "color scheme" that is applied to the UI? And does it intend to share the default color scheme with Calypso, or does it define its own?

Color scheme defines approx 15-20 colors, all selected from the `color-studio` palette, while the whole `color-studio` has 170+ colors.

Comment (folletto) [reply postcss.config.js]: Not explicitly. I think the code we have already there is the closest thing we have to a color scheme.

For sure, we won't use all the colors in the `color-studio`, but right now there isn't a logical subset that has been extracted — again, except what's already there in code.

**Pull request merged**
